File: ios/Classes/AdropAdsFlutterPlugin.swift
import Flutter
import UIKit
import WebKit
import AdropAds

public class AdropAdsFlutterPlugin: NSObject, FlutterPlugin {
    private var bannerManager: AdropBannerManager?
    private var adropAdManager : AdropAdManager?
    private var messenger: FlutterBinaryMessenger?
    private var nativeViewFactory: AdropNativeAdViewFactory?
    private let consentManager = FlutterAdropConsentManager()

    private let ModuleError = FlutterError(
        code: "ERROR_CODE_INTERNAL",
        message: "Module is undefined",
        details: "Expected adType enum index larger than 0"
    )

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = AdropAdsFlutterPlugin()
        let messenger = registrar.messenger()
        instance.messenger = messenger

        let adManager = AdropAdManager(messenger: messenger)
        instance.adropAdManager = adManager

        let invokeChannel = FlutterMethodChannel(name: AdropChannel.invokeChannel, binaryMessenger: messenger)
        let bannerManager = AdropBannerManager(messenger: messenger)
        registrar.addMethodCallDelegate(instance, channel: invokeChannel)

        instance.bannerManager = bannerManager
        registrar.register(AdropBannerViewFactory(messenger: messenger, bannerManager: bannerManager), withId: AdropChannel.bannerEventListenerChannel)

        instance.nativeViewFactory = AdropNativeAdViewFactory(messenger: messenger, adManager: adManager)
        guard let nativeViewFactory = instance.nativeViewFactory else { return }

        registrar.register(nativeViewFactory, withId: AdropChannel.nativeEventListenerChannel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case AdropMethod.INITIALIZE:
            let args = call.arguments as? [String: Any?]
            let production = args?["production"] as? Bool ?? false
            let targetCountries = args?["targetCountries"] as? [String]
            let useInAppBrowser = args?["useInAppBrowser"] as? Bool ?? false
            Adrop.initialize(production: production, useInAppBrowser: useInAppBrowser, targetCountries: targetCountries)
            result(nil)
        case AdropMethod.SET_UID:
            let uid = (call.arguments as? [String: Any?])?["uid"] as? String ?? ""
            if (uid.isEmpty) {
                result(FlutterError(
                    code: "ERROR_CODE_INTERNAL",
                    message: "Invalid uid",
                    details: "Expected not empty string"
                ))
                return
            }

            Adrop.setUID(uid)
            result(nil)
        case AdropMethod.SET_THEME:
            let theme = (call.arguments as? [String: Any?])?["theme"] as? String ?? "auto"
            let converted: AdropTheme
            switch theme.lowercased() {
            case "light":
                converted = .light
            case "dark":
                converted = .dark
            default:
                converted = .auto
            }

            Adrop.setTheme(converted)
            result(nil)
        case AdropMethod.SET_PROPERTY:
            let key = (call.arguments as? [String: Any?])?["key"] as? String ?? ""
            let value = (call.arguments as? [String: Any?])?["value"] as? Any ?? [:]
            AdropMetrics.setProperty(key: key, value: convertToEncodable(value))
            result(nil)
        case AdropMethod.GET_PROPERTY:
            result(AdropMetrics.properties())
        case AdropMethod.LOG_EVENT:
            let name = (call.arguments as? [String: Any?])?["name"] as? String ?? ""
            let params = (call.arguments as? [String: Any?])?["params"] as? [String:Any] ?? [:]
            var encodableParams: [String: Encodable] = [:]

            for (key, value) in params {
                if let encodableValue = convertToEncodable(value) {
                    encodableParams[key] = encodableValue
                }
            }
            AdropMetrics.logEvent(name: name, params: encodableParams)
            result(nil)

        case AdropMethod.SEND_EVENT:
            let name = (call.arguments as? [String: Any?])?["name"] as? String ?? ""
            let params = (call.arguments as? [String: Any?])?["params"] as? [String: Any]

            AdropMetrics.sendEvent(name: name, params: params)
            result(nil)

        case AdropMethod.LOAD_BANNER:
            let args = call.arguments as? [String: Any?]
            let unitId = args?["unitId"] as? String ?? ""
            let requestId = args?["requestId"] as? String ?? ""
            let width = args?["width"] as? CGFloat
            let height = args?["height"] as? CGFloat

            bannerManager?.load(unitId: unitId, requestId: requestId, width: width, height: height)
            result(nil)
        case AdropMethod.DISPOSE_BANNER:
            let unitId = (call.arguments as? [String: Any?])?["unitId"] as? String ?? ""
            let requestId = (call.arguments as? [String: Any?])?["requestId"] as? String ?? ""

            bannerManager?.destroy(unitId: unitId, requestId: requestId)
            result(nil)
        case AdropMethod.PLAY_BANNER:
            let unitId = (call.arguments as? [String: Any?])?["unitId"] as? String ?? ""
            let requestId = (call.arguments as? [String: Any?])?["requestId"] as? String ?? ""

            bannerManager?.play(unitId: unitId, requestId: requestId)
            result(nil)
        case AdropMethod.PAUSE_BANNER:
            let unitId = (call.arguments as? [String: Any?])?["unitId"] as? String ?? ""
            let requestId = (call.arguments as? [String: Any?])?["requestId"] as? String ?? ""

            bannerManager?.pause(unitId: unitId, requestId: requestId)
            result(nil)
        case AdropMethod.LOAD_AD:
            let adTypeIndex = (call.arguments as? [String: Any?])?["adType"] as? Int ?? 0
            if (adTypeIndex == AdType.allCases.firstIndex(of: .undefined)!) {
                result(ModuleError)
                return
            }
            let loadArgs = call.arguments as? [String: Any?]
            let userId = loadArgs?["userId"] as? String
            let customData = loadArgs?["customData"] as? String
            let ssvOptions: AdropServerSideVerificationOptions? = (userId != nil || customData != nil)
                ? AdropServerSideVerificationOptions(userId: userId, customData: customData) : nil

            adropAdManager?.load(
                adType: AdType.allCases[adTypeIndex],
                unitId: loadArgs?["unitId"] as? String ?? "",
                requestId: loadArgs?["requestId"] as? String ?? "",
                useCustomClick: loadArgs?["useCustomClick"] as? Bool ?? false,
                ssvOptions: ssvOptions
            )
            result(nil)
        case AdropMethod.SHOW_AD:
            let adTypeIndex = (call.arguments as? [String: Any?])?["adType"] as? Int ?? 0
            if (adTypeIndex == AdType.allCases.firstIndex(of: .undefined)!) {
                result(ModuleError)
                return
            }
            adropAdManager?.show(
                adType: AdType.allCases[adTypeIndex],
                requestId: (call.arguments as? [String: Any?])?["requestId"] as? String ?? ""
            )
            result(nil)
        case AdropMethod.CLOSE_AD:
            let adTypeIndex = (call.arguments as? [String: Any?])?["adType"] as? Int ?? 0
            if (adTypeIndex == AdType.allCases.firstIndex(of: .undefined)!) {
                result(ModuleError)
                return
            }
            adropAdManager?.close(
                adType: AdType.allCases[adTypeIndex],
                requestId: (call.arguments as? [String: Any?])?["requestId"] as? String ?? ""
            )
            result(nil)
        case AdropMethod.CUSTOMIZE_AD:
            let adTypeIndex = (call.arguments as? [String: Any?])?["adType"] as? Int ?? 0
            if (adTypeIndex == AdType.allCases.firstIndex(of: .undefined)!) {
                result(ModuleError)
                return
            }
            adropAdManager?.customize(
                adType: AdType.allCases[adTypeIndex],
                requestId:(call.arguments as? [String: Any?])?["requestId"] as? String ?? "",
                data: (call.arguments as? [String: Any?])?["data"] as? [String:Any] ?? [:])
            result(nil)
        case AdropMethod.DISPOSE_AD:
            let adTypeIndex = (call.arguments as? [String: Any?])?["adType"] as? Int ?? 0
            if (adTypeIndex == AdType.allCases.firstIndex(of: .undefined)!) {
                result(ModuleError)
                return
            }
            adropAdManager?.destroy(
                adType: AdType.allCases[adTypeIndex],
                requestId: (call.arguments as? [String: Any?])?["requestId"] as? String ?? ""
            )
            result(nil)
        case AdropMethod.PERFORM_CLICK:
            let requestId = (call.arguments as? [String: Any?])?["requestId"] as? String ?? ""
            nativeViewFactory?.performClick(requestId)
            result(nil)

        case AdropMethod.REQUEST_CONSENT_INFO_UPDATE:
            consentManager.requestConsentInfoUpdate(result: result)

        case AdropMethod.GET_CONSENT_STATUS:
            consentManager.getConsentStatus(result: result)

        case AdropMethod.CAN_REQUEST_ADS:
            consentManager.canRequestAds(result: result)

        case AdropMethod.RESET_CONSENT:
            consentManager.reset(result: result)

        case AdropMethod.SET_CONSENT_DEBUG_SETTINGS:
            let geographyValue = (call.arguments as? [String: Any?])?["geography"] as? Int ?? 0
            consentManager.setDebugSettings(geographyValue: geographyValue, result: result)

        case AdropMethod.REGISTER_WEB_VIEW:
            let webViewId = (call.arguments as? [String: Any?])?["webViewId"] as? Int64 ?? -1
            if webViewId >= 0 {
                registerWebView(identifier: webViewId)
            }
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func registerWebView(identifier: Int64) {
        guard let registry = UIApplication.shared.delegate as? FlutterPluginRegistry else { return }

        guard let externalApiClass = NSClassFromString(
            "FWFWebViewFlutterWKWebViewExternalAPI"
        ) as? NSObject.Type else { return }

        let selector = NSSelectorFromString("webViewForIdentifier:withPluginRegistry:")
        guard externalApiClass.responds(to: selector) else { return }
        guard let method = class_getClassMethod(externalApiClass, selector) else { return }

        typealias WebViewLookup = @convention(c) (AnyObject, Selector, Int64, AnyObject) -> WKWebView?
        let imp = method_getImplementation(method)
        let function = unsafeBitCast(imp, to: WebViewLookup.self)

        if let webView = function(externalApiClass, selector, identifier, registry as AnyObject) {
            Adrop.registerWebView(webView)
        }
    }

    private func convertToEncodable(_ value: Any) -> Encodable? {
        switch value {
        case let numberValue as NSNumber:
            if CFGetTypeID(numberValue) == CFBooleanGetTypeID() {
                return numberValue.boolValue
            } else if CFGetTypeID(numberValue) == CFNumberGetTypeID() {
                if numberValue is Int {
                    return  numberValue.intValue
                } else if numberValue is Double || numberValue is Float {
                    return numberValue.doubleValue
                } else {
                    return nil
                }
            } else {
                return nil
            }
        case let boolValue as Bool:
            return boolValue
        case let stringValue as String:
            return stringValue
        case let intValue as Int:
            return intValue
        case let doubleValue as Double:
            return doubleValue
        case _ as [Any]:
            // array not supported
            return nil
        case _ as [String: Any]:
            // dictionary not supported
            return nil
        default:
            if let encodableValue = value as? Encodable {
                return encodableValue
            } else {
                return nil
            }
        }
    }

    func covertAnyToEncodable(_ value: Any) -> Encodable? {
        return convertToEncodable(value)
    }
}

File: android/src/main/kotlin/io/adrop/adrop_ads/AdropAdsFlutterPlugin.kt
package io.adrop.adrop_ads

import android.app.Activity
import android.app.Application
import android.content.Context
import androidx.annotation.NonNull
import io.adrop.adrop_ads.banner.AdropBannerManager
import io.adrop.adrop_ads.banner.AdropBannerViewFactory
import io.adrop.adrop_ads.bridge.AdropChannel
import io.adrop.adrop_ads.bridge.AdropError
import io.adrop.adrop_ads.bridge.AdropMethod
import io.adrop.adrop_ads.metrics.FlutterAdropMetrics
import io.adrop.adrop_ads.native.AdropNativeAdViewFactory
import io.adrop.ads.Adrop
import io.adrop.ads.metrics.AdropEventParam
import io.adrop.ads.metrics.AdropMetrics
import io.adrop.ads.model.AdropErrorCode
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result


class AdropAdsFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private var activity: Activity? = null
    private var context: Context? = null

    private lateinit var invokeChannel: MethodChannel
    private lateinit var bannerManager: AdropBannerManager
    private lateinit var messenger: BinaryMessenger
    private val adManager = AdropAdManager()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        messenger = flutterPluginBinding.binaryMessenger
        invokeChannel = MethodChannel(messenger, AdropChannel.INVOKE_CHANNEL)
        invokeChannel.setMethodCallHandler(this)
        bannerManager = AdropBannerManager(context, messenger)

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            AdropChannel.BANNER_EVENT_LISTENER_CHANNEL,
            AdropBannerViewFactory(bannerManager),
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            AdropChannel.NATIVE_EVENT_LISTENER_CHANNEL,
            AdropNativeAdViewFactory(adManager),
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        invokeChannel.setMethodCallHandler(null)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        try {
            when (call.method) {
                AdropMethod.INITIALIZE -> {
                    val production = call.argument("production") as Boolean? ?: false
                    val targetCountries = call.argument("targetCountries") as List<String>?
                    initialize(production, targetCountries?.toTypedArray())
                    result.success(null)
                }
                AdropMethod.SET_UID -> {
                    val uid = call.argument("uid") as String? ?: ""
                    if (uid.isEmpty()) {
                        result.error(
                            AdropErrorCode.ERROR_CODE_INTERNAL.name,
                            "Invalid uid",
                            "Expected not empty string"
                        )
                        return
                    }

                    Adrop.setUID(uid)
                    result.success(null)
                }

                AdropMethod.SET_PROPERTY -> {
                    val key = call.argument("key") as String? ?: ""
                    val value = call.argument("value") as Any? ?: arrayOf<Int>()

                    FlutterAdropMetrics.setProperty(key, value)

                    result.success(null)
                }

                AdropMethod.GET_PROPERTY -> {
                    result.success(FlutterAdropMetrics.properties())
                }

                AdropMethod.LOG_EVENT -> {
                    val name = call.argument("name") as String? ?: ""
                    val mapValue = call.argument("params") as? Map<String, Any> ?: HashMap()

                    val builder = AdropEventParam.Builder()
                    mapValue.keys.forEach {
                        when (val value = mapValue[it]) {
                            is String -> builder.putString(it, value)
                            is Int -> builder.putInt(it, value)
                            is Float -> builder.putFloat(it, value)
                            is Double -> builder.putFloat(it, value.toFloat())
                            is Boolean -> builder.putBoolean(it, value)
                            else -> {}
                        }
                    }

                    AdropMetrics.logEvent(name, builder.build())
                    result.success(null)
                }

                AdropMethod.LOAD_BANNER -> {
                    val unitId = call.argument("unitId") as String? ?: ""
                    val requestId = call.argument("requestId") as String? ?: ""
                    bannerManager.load(unitId, requestId)
                    result.success(null)
                }

                AdropMethod.DISPOSE_BANNER -> {
                    val unitId = call.argument("unitId") as String? ?: ""
                    val requestId = call.argument("requestId") as String? ?: ""
                    bannerManager.destroy(unitId, requestId)
                    result.success(null)
                }

                AdropMethod.LOAD_AD -> {
                    if (context == null) {
                        result.error(AdropErrorCode.ERROR_CODE_INITIALIZE.name, "method call received before context initialized", null)
                        return
                    }
                    val adTypeIndex = call.argument("adType") as Int? ?: 0
                    if (adTypeIndex == AdType.Undefined.ordinal) {
                        result.error(AdropErrorCode.ERROR_CODE_INTERNAL.name, "AdType is undefined", "Expected adType enum index larger than 0")
                        return
                    }
                    adManager.load(
                        context!!,
                        AdType.values()[adTypeIndex],
                        call.argument("unitId") as String? ?: "",
                        call.argument("requestId") as String? ?: "",
                        messenger
                    )
                    result.success(null)
                }

                AdropMethod.SHOW_AD -> {
                    if (activity == null) {
                        result.error(AdropErrorCode.ERROR_CODE_INITIALIZE.name, "method call received before activity initialized", null)
                        return
                    }
                    val adTypeIndex = call.argument("adType") as Int? ?: 0
                    if (adTypeIndex == AdType.Undefined.ordinal) {
                        result.error(AdropErrorCode.ERROR_CODE_INTERNAL.name, "AdType is undefined", "Expected adType enum index larger than 0")
                        return
                    }

                    val requestId = call.argument("requestId") as String? ?: ""

                    adManager.show(
                        AdType.values()[adTypeIndex],
                        requestId,
                        activity!!
                    )
                    result.success(null)
                }

                AdropMethod.CUSTOMIZE_AD -> {
                    val adTypeIndex = call.argument("adType") as Int? ?: 0
                    if (adTypeIndex == AdType.Undefined.ordinal) {
                        result.error(AdropErrorCode.ERROR_CODE_INTERNAL.name, "AdType is undefined", "Expected adType enum index larger than 0")
                        return
                    }

                    adManager.customize(
                        AdType.values()[adTypeIndex],
                        call.argument("requestId") as String? ?: "",
                        call.argument("data") as? Map<String, Any> ?: HashMap()
                        )
                    result.success(null)
                }

                AdropMethod.DISPOSE_AD -> {
                    val adTypeIndex = call.argument("adType") as Int? ?: 0
                    if (adTypeIndex == AdType.Undefined.ordinal) {
                        result.error(AdropErrorCode.ERROR_CODE_INTERNAL.name, "AdType is undefined", "Expected adType enum index larger than 0")
                        return
                    }
                    adManager.destroy(
                        AdType.values()[adTypeIndex],
                        call.argument("requestId") as String? ?: ""
                    )
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        } catch (e: AdropError) {
            result.error(e.code, e.message, null)
        } catch (e: Exception) {
            e.printStackTrace()
        } catch (e: Error) {
            e.printStackTrace()
        }
    }

    @Throws(AdropError::class)
    private fun initialize(production: Boolean?, targetCountries: Array<String>? = null) {
        val context = context ?: return
        if (context is Application) {
            Adrop.initialize(context, production ?: false, targetCountries ?: arrayOf())
            return
        } else {
            throw AdropError(AdropErrorCode.ERROR_CODE_INITIALIZE)
        }
    }

    /** ActivityAware **/

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        this.activity = binding.activity
        this.context = binding.activity.applicationContext
    }

    override fun onDetachedFromActivityForConfigChanges() {
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        this.activity = binding.activity
        this.context = binding.activity.applicationContext
    }

    override fun onDetachedFromActivity() {
    }
}
